### PR TITLE
CBL-4009: Implement FullTextIndexExpression per doc changes with tests

### PR DESCRIPTION
**Pull request by bmeike**
CBL-4009: Implement FullTextIndexExpression per doc changes with tests
CBL-3352: Add doc discussing Notifier/Token mem management

Also: 
 - Add ':' and ';' to the list of illegal characters tested in Collection creation
 - Refactor Expresssion.FunctionExpression constructor to be varadic.

### CODE REVIEW

**Comment by bmeike** [comment]
Force pushes to clean up some stuff.

File: common/main/java/com/couchbase/lite/Expression.java [hunk 1]
```diff
@@ -329,6 +328,65 @@ Object asJSON() {
         }
     }
 
+    //---------------------------------------------
```
Comment (bmeike) [inline common/main/java/com/couchbase/lite/Expression.java:331]: This is the compilable Expression representing an index

File: common/main/java/com/couchbase/lite/Expression.java [hunk 2]
```diff
@@ -488,6 +546,21 @@ public static Expression not(@NonNull Expression expression) {
         return negated(expression);
     }
 
+    /**
```
Comment (bmeike) [inline common/main/java/com/couchbase/lite/Expression.java:549]: New API per spec

File: common/main/java/com/couchbase/lite/FullTextFunction.java [hunk 3]
```diff
@@ -33,18 +31,31 @@ public final class FullTextFunction {
     private FullTextFunction() { }
 
     /**
```
Comment (bmeike) [inline common/main/java/com/couchbase/lite/FullTextFunction.java:33]: New API, per spec. Old API deprecated

File: common/main/java/com/couchbase/lite/IndexExpression.java [hunk 4]
```diff
@@ -0,0 +1,19 @@
+//
```
Comment (bmeike) [inline common/main/java/com/couchbase/lite/IndexExpression.java:1]: New type, per spec

File: common/main/java/com/couchbase/lite/internal/listener/ChangeNotifier.java [hunk 5]
```diff
@@ -27,7 +27,27 @@
 import com.couchbase.lite.ListenerToken;
 import com.couchbase.lite.internal.utils.Fn;
```
Comment (bmeike) [inline common/main/java/com/couchbase/lite/internal/listener/ChangeNotifier.java:29]: Documentation for CBL-3352

File: common/test/java/com/couchbase/lite/BaseDbTest.kt [hunk 6]
```diff
@@ -45,10 +45,10 @@ val Scope.collectionCount
     get() = this.collections.size
 
 fun CountDownLatch.stdWait(): Boolean {
-    try {
-        return await(BaseTest.STD_TIMEOUT_SEC, TimeUnit.SECONDS)
+    return try {
```
Comment (bmeike) [inline common/test/java/com/couchbase/lite/BaseDbTest.kt:48]: Cleanup

File: common/test/java/com/couchbase/lite/DbCollectionsTest.kt [hunk 7]
```diff
@@ -26,7 +26,8 @@ import org.junit.Test
 
 class DbCollectionsTest : BaseDbTest() {
     private val invalidChars = charArrayOf(
```
Comment (bmeike) [inline common/test/java/com/couchbase/lite/DbCollectionsTest.kt:28]: Add ';', ':' as illegal characters for Collection names

File: common/main/java/com/couchbase/lite/AbstractFunction.java [hunk 8]
```diff
@@ -484,13 +482,11 @@ public static Expression contains(@NonNull Expression operand, @NonNull Expressi
 
     @NonNull
     private static Expression.FunctionExpression sexpr(@NonNull String expr, @NonNull Expression operand) {
```
Comment (bmeike) [inline common/main/java/com/couchbase/lite/AbstractFunction.java:484]: Refactor Expression.FunctionExpression to be varadic

File: common/main/java/com/couchbase/lite/Collection.java [hunk 9]
```diff
@@ -491,7 +491,7 @@ public boolean equals(Object o) {
 
 
     @NonNull
```
Comment (bmeike) [inline common/main/java/com/couchbase/lite/Collection.java:493]: Code cleanup: this methods should not be visible in the API

File: common/main/java/com/couchbase/lite/Collection.java [hunk 10]
```diff
@@ -458,7 +458,7 @@ public void deleteIndex(String name) throws CouchbaseLiteException {
     @Override
     public void close() {
         synchronized (getDbLock()) {
-            closeCollectionChangNotifierLocked();
```
Comment (bmeike) [inline common/main/java/com/couchbase/lite/Collection.java:461]: Cleanup: Name misspelled.

File: common/main/java/com/couchbase/lite/Collection.java [hunk 11]
```diff
@@ -877,7 +875,7 @@ else if (!document.isEmpty()) {
         }
     }
```
Comment (bmeike) [inline common/main/java/com/couchbase/lite/Collection.java:877]: Cleanup: misspelled name

File: common/main/java/com/couchbase/lite/internal/listener/ChangeNotifier.java [hunk 12]
```diff
@@ -27,7 +27,27 @@
...
+// little memory.  Notifiers, however, hold references to Core companion
+// objects.  When the last listener is removed (removeListener returns true)
+// the notifier should be closed and the companion objects freed.
+// This moots the concerning scenario in which client code simply drops
+// the token on the floor.  If it is never released, it and the notifier
+// that contains it will not be released: the Core objects will be retained.
+// Having the tokens free themselves in their finalizers is not a good approach.
+// If you do that, as soon as the token is GCed, the associated listener will
+// stop listening, surely surprising to the client code.
+// On the other hand, once the replicator is freed, all of its notifiers are
```
Comment (pasin) [inline common/main/java/com/couchbase/lite/internal/listener/ChangeNotifier.java:47]: Assume it means that when the objects that keep the notifiers, not just the replicator, are freed, the underlining Core's Observer will be freed?

File: common/main/java/com/couchbase/lite/CollectionChangeNotifier.java [hunk 13]
```diff
@@ -29,7 +29,7 @@
 import com.couchbase.lite.internal.utils.Preconditions;
 
 
-// Not thread safe...
```
Comment (bmeike) [inline common/main/java/com/couchbase/lite/CollectionChangeNotifier.java:32]: Remove inaccurate comment

**Review by pasin — CHANGES_REQUESTED** [review]

File: common/main/java/com/couchbase/lite/DocumentChangeNotifier.java [hunk 14]
```diff
@@ -23,7 +23,7 @@
 import com.couchbase.lite.internal.listener.ChangeNotifier;
 import com.couchbase.lite.internal.utils.Fn;
 
-
```
Comment (bmeike) [inline common/main/java/com/couchbase/lite/DocumentChangeNotifier.java:26]: Doc for CBL-3352

File: common/main/java/com/couchbase/lite/Expression.java [hunk 15]
```diff
@@ -35,7 +35,7 @@
 /**
  * The expression used in constructing a query.
  */
-@SuppressWarnings("PMD.TooManyMethods")
```
Comment (bmeike) [inline common/main/java/com/couchbase/lite/Expression.java:38]: API is per-spec.  PMD can take a seat.

File: common/main/java/com/couchbase/lite/Expression.java [hunk 16]
```diff
@@ -126,7 +126,6 @@ Object asJSON() {
     //---------------------------------------------
     // Binary Expression
     //---------------------------------------------
-    @SuppressWarnings("PMD.FieldNamingConventions")
```
Comment (bmeike) [inline common/main/java/com/couchbase/lite/Expression.java:129]: Remove unnecessary warning suppression

File: common/main/java/com/couchbase/lite/Expression.java [hunk 17]
```diff
@@ -314,9 +313,9 @@ static final class FunctionExpression extends Expression {
         @NonNull
         private final List<Expression> params;
 
-        FunctionExpression(@NonNull String func, @NonNull List<Expression> params) {
```
Comment (bmeike) [inline common/main/java/com/couchbase/lite/Expression.java:317]: Make Expression.FunctionExpression varadic

File: common/main/java/com/couchbase/lite/IndexExpression.java [hunk 18]
```diff
@@ -0,0 +1,19 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+package com.couchbase.lite;
+
+
+interface IndexExpression { }
```
[the remark this answers is not in the document]
Comment (bmeike) [reply common/main/java/com/couchbase/lite/IndexExpression.java:19]: Whoops.  Yes.

File: common/main/java/com/couchbase/lite/internal/listener/ChangeNotifier.java (hunk 12 above)
Comment (bmeike) [reply common/main/java/com/couchbase/lite/internal/listener/ChangeNotifier.java:47]: No.  In order to free the Core observer, one of two things must happen: either a call to token.remove() or the Replicator object becomes unreachable (in the Java sense)

File: common/main/java/com/couchbase/lite/Collection.java [hunk 19]
```diff
@@ -704,7 +702,7 @@ ListenerToken addDocumentChangeListenerLocked(
             docChangeNotifiers.put(docID, docNotifier);
             if (isOpen()) {
                 try { docNotifier.start(this::scheduleImmediateOnPostExecutor); }
-                // !!! Revisit this: there is no programatic way for client code
+                // ??? Revisit this: there is no programatic way for client code
                 // to know that the listener has failed.
                 catch (CouchbaseLiteException e) {
                     Log.d(LogDomain.LISTENER, "Listener failed", e);
```
[the remark this answers is not in the document]
Comment (pasin) [reply common/main/java/com/couchbase/lite/Collection.java:708]: Are you creating a ticket or you want me to create a ticket?

**Review by pasin — APPROVED** [review]

**Pull request merged**